File: v2/src/main/java/com/api/v2/people/services/PersonUpdatingServiceImpl.java
package com.api.v2.people.services;

import com.api.v2.people.domain.PersonAuditTrail;
import com.api.v2.people.domain.PersonAuditTrailRepository;
import com.api.v2.people.domain.PersonRepository;
import com.api.v2.people.domain.exposed.Person;
import com.api.v2.people.requests.PersonUpdatingDto;
import com.api.v2.people.services.exposed.PersonUpdatingService;
import jakarta.validation.Valid;
import jakarta.validation.constraints.NotNull;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

@Service
@RequiredArgsConstructor
public class PersonUpdatingServiceImpl implements PersonUpdatingService {

    private final PersonAuditTrailRepository personAuditTrailRepository;
    private final PersonRepository personRepository;

    @Override
    public Mono<Person> update(@NotNull Person oldPerson, @Valid PersonUpdatingDto updatingDto) {
        return Mono.defer(() -> {
            PersonAuditTrail auditTrail = PersonAuditTrail.of(oldPerson);
            return personAuditTrailRepository
                    .save(auditTrail)
                    .flatMap(ignored -> {
                        Person updatedPerson = Person.of(updatingDto, oldPerson.ssn());
                        return personRepository.save(updatedPerson);
                    });
        });
    }
}

File: v2/src/main/kotlin/com/api/v2/people/services/PersonUpdatingServiceImpl.kt
package com.api.v2.people.services

import com.api.v2.people.domain.PersonAuditTrail
import com.api.v2.people.domain.PersonAuditTrailRepository
import com.api.v2.people.domain.PersonRepository
import com.api.v2.people.domain.exposed.Person
import com.api.v2.people.requests.PersonUpdatingDTO
import com.api.v2.people.services.exposed.PersonUpdatingService
import jakarta.validation.Valid
import jakarta.validation.constraints.NotNull
import org.springframework.stereotype.Service

@Service
class PersonUpdatingServiceImpl(
    private val auditTrailRepository: PersonAuditTrailRepository,
    private val repository: PersonRepository
) : PersonUpdatingService {

    override suspend fun update(
        person: @NotNull Person,
        updatingDTO: @Valid PersonUpdatingDTO
    ): Person {
        val auditTrail = PersonAuditTrail.of(person)
        val savedAuditTrail = auditTrailRepository.save(auditTrail)
        person.update(updatingDTO)
        return repository.save(person)
    }
}